cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -1844,10 +1844,14 @@ macro(build_gtest)
   set(GTEST_VENDORED TRUE)
   set(GTEST_CMAKE_CXX_FLAGS ${EP_CXX_FLAGS})
 
-  if(CMAKE_BUILD_TYPE MATCHES DEBUG)
-    set(CMAKE_GTEST_DEBUG_EXTENSION "d")
+  if(NOT CMAKE_GENERATOR STREQUAL Xcode)
+    if(CMAKE_BUILD_TYPE MATCHES DEBUG)
+      set(CMAKE_GTEST_DEBUG_EXTENSION "d")
+    else()
+      set(CMAKE_GTEST_DEBUG_EXTENSION "")
+    endif()
   else()
-    set(CMAKE_GTEST_DEBUG_EXTENSION "")
+    set(CMAKE_GTEST_DEBUG_EXTENSION "d")
```

Comment:
I don't understand, why is it always "d" for XCode?

Reply:
As far as I can tell, it's because Xcode selects whether to build in debug or release mode after configuration, so if it's not set then it will succeed building in Release but not in Debug because it'll try to find `libgtest.dylib` instead of `libgtestd.dylib`. Without this change, if I specified `-DCMAKE_BUILD_TYPE=Debug`, Xcode would be able to build in debug mode but not release mode.

Reply:
Do you mean it always builds GTest in debug mode?

Reply:
I think it'll build GTest the correct way, but it won't name it properly.

Reply:
Hmm...I get this error in Windows if I do `cmake --build .` instead of `cmake --config Debug --build .`

Coming from Linux this is confusing since I never have to specify `--config` but perhaps the issue here is related?

Reply:
Could be - what do you use to build on Windows? Visual Studio?

Reply:
Both visual studio and command line (with visual studio generator)

Reply:
Ah yeah then that might be the same problem. Visual Studio probably also sets the build type independently of configuration.

Reply:
Any chance you can test out this fix on Windows?

Reply:
How about using generator expression https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html ? We can use generator expression in `ExternalProject_Add`'s arguments: https://cmake.org/cmake/help/latest/module/ExternalProject.html

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index ff48ac7ee5..44cab1d6b1 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -1849,11 +1849,7 @@ macro(build_gtest)
   set(GTEST_VENDORED TRUE)
   set(GTEST_CMAKE_CXX_FLAGS ${EP_CXX_FLAGS})
 
-  if(CMAKE_BUILD_TYPE MATCHES DEBUG)
-    set(CMAKE_GTEST_DEBUG_EXTENSION "d")
-  else()
-    set(CMAKE_GTEST_DEBUG_EXTENSION "")
-  endif()
+  set(CMAKE_GTEST_DEBUG_EXTENSION "$<$<CONFIG:Debug>,d>")
 
   if(APPLE)
     set(GTEST_CMAKE_CXX_FLAGS ${GTEST_CMAKE_CXX_FLAGS} -DGTEST_USE_OWN_TR1_TUPLE=1
@@ -1894,9 +1890,9 @@ macro(build_gtest)
   set(GTEST_CMAKE_ARGS
       ${EP_COMMON_TOOLCHAIN}
       -DBUILD_SHARED_LIBS=ON
-      -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+      -DCMAKE_BUILD_TYPE=$<CONFIG>
       -DCMAKE_CXX_FLAGS=${GTEST_CMAKE_CXX_FLAGS}
-      -DCMAKE_CXX_FLAGS_${UPPERCASE_BUILD_TYPE}=${GTEST_CMAKE_CXX_FLAGS}
+      -DCMAKE_CXX_FLAGS_$<UPPER_CASE:$<CONFIG>>=${GTEST_CMAKE_CXX_FLAGS}
       -DCMAKE_INSTALL_LIBDIR=lib
       -DCMAKE_INSTALL_NAME_DIR=${GTEST_INSTALL_NAME_DIR}
       -DCMAKE_INSTALL_PREFIX=${GTEST_PREFIX}
```